### PR TITLE
OCPNODE-2213: Add support for Recursive Read-only (RRO) mounts

### PR DESCRIPTION
**Pull request by kwilczynski**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Currently, the read-only option set for the host mounts does not correctly propagate recursively from the top level of the given mount point. As such, any sub-mounts that happen to be part of the tree would retain their current options, and should it be a read/write type mount, then it could be modified without any restriction. This is often both unexpected and undesirable.

Thus, add support for so-called Recursive Read-only (RRO) mounts. This allows a mount point to propagate the read-only option through the tree, marking the entire mount point read-only.

Support for the RRO mounts has already been available in the runtimes, such as **crun** and **runc**, for some time now.

> [!NOTE] 
> Support of the Recursive Read-only (RRO) mounts requires:
>
> - For Linux, kernel version 5.12
> - Kubernetes v1.30🔹
> - cri-tools v1.30
> - crun v1.4
> - runc v1.1
>
> 🔹⚠️ Feature gate called **RecursiveReadOnlyMounts** needs to be enabled given the _Alpha_ status of the feature.

Related to:

- https://github.com/kubernetes/enhancements/issues/3857
- https://github.com/kubernetes/enhancements/pull/3858
- https://github.com/kubernetes/kubernetes/pull/123272
- https://github.com/kubernetes-sigs/cri-tools/pull/1344
- https://github.com/opencontainers/runc/pull/3272
- https://github.com/containers/crun/pull/778

#### Which issue(s) this PR fixes:

A temporary workaround:

- https://github.com/cri-o/cri-o/pull/7828

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
Add support for the Recursive Read-only (RRO) mounts following runtimes and kernel support to enable read-only mounts to be recursively read-only.
```

### CODE REVIEW

File: server/server_linux.go
Comment (haircommander) [inline server/server_linux.go:168]: I am fine with this for now though eventually it would be better if we found a way to check the feature without referencing the version, as it could have been backported.

Comment (haircommander) [reply server/server_linux.go:168]: ah nevermind I see the TODO below

File: server/server_linux.go
Comment (haircommander) [reply server/server_linux.go:150]: BTW I think I'd prefer this inline below, rather than saving the functor and calling it below. I don't think we save on indentation or code readability as-is

File: server/runtime_status.go
Comment (haircommander) [inline server/runtime_status.go:55]: we call this a lot, potentially unnecessarily. I think we should check this in `initializeRuntimeFeatures()` in `pkg/config/config.go` and if the kernel doesn't support the flag, then remove it from the features structure. Then, from that point on, we can just ask the runtime whether it's supported or not

File: server/server_linux.go
Comment (kwilczynski) [reply server/server_linux.go:150]: Sure! Will do.

File: server/container_create_linux.go
Comment (haircommander) [inline server/container_create_linux.go:1166]: we can do this with a flat if/else:
```
if rro && ro {
// rro
} else if rro
// err
} else if ro {
// ro
} else {
// rw
}
```
which I think would be easier to read

File: server/server_linux.go
Comment (haircommander) [reply server/server_linux.go:168]: yeah makes sense to me

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go:1166]: Sure. I can do it like that.

For the record, I personally find nested if-else-if blocks harder to read.

Comment (haircommander) [reply server/container_create_linux.go:1166]: yeah It would be best as a switch but we have two variables to check :(

File: server/runtime_status.go
Comment (kwilczynski) [reply server/runtime_status.go:55]: I see what you mean. I think we can plumb this like that. Let me see. :+1:

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go]: Done.

File: server/server_linux.go
Comment (kwilczynski) [reply server/server_linux.go]: Done.

File: server/runtime_status.go
Comment (kwilczynski) [reply server/runtime_status.go:55]: We unmarshal the JSON atop the following:

_(from [opencontainers/runtime-spec/specs-go/features/features.go](https://github.com/opencontainers/runtime-spec/blob/main/specs-go/features/features.go#L5-L33)):_

```go
type Features struct {
	// OCIVersionMin is the minimum OCI Runtime Spec version recognized by the runtime, e.g., "1.0.0".
	OCIVersionMin string `json:"ociVersionMin,omitempty"`

	// OCIVersionMax is the maximum OCI Runtime Spec version recognized by the runtime, e.g., "1.0.2-dev".
	OCIVersionMax string `json:"ociVersionMax,omitempty"`

	// Hooks is the list of the recognized hook names, e.g., "createRuntime".
	// Nil value means "unknown", not "no support for any hook".
	Hooks []string `json:"hooks,omitempty"`

	// MountOptions is the list of the recognized mount options, e.g., "ro".
	// Nil value means "unknown", not "no support for any mount option".
	// This list does not contain filesystem-specific options passed to mount(2) syscall as (const void *).
	MountOptions []string `json:"mountOptions,omitempty"`

	// Linux is specific to Linux.
	Linux *Linux `json:"linux,omitempty"`

	// Annotations contains implementation-specific annotation strings,
	// such as the implementation version, and third-party extensions.
	Annotations map[string]string `json:"annotations,omitempty"`

	// PotentiallyUnsafeConfigAnnotations the list of the potential unsafe annotations
	// that may appear in `config.json`.
	//
	// A value that ends with "." is interpreted as a prefix of annotations.
	PotentiallyUnsafeConfigAnnotations []string `json:"potentiallyUnsafeConfigAnnotations,omitempty"`
}
```

This is not the same as the CRI API type, which I think you had in mind here?

File: server/server_linux.go
Comment (kwilczynski) [reply server/server_linux.go]: Done.

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go]: I am going to rewrite this using a switch statement, since linter picks up on it, per:

```golang
server/container_create_linux.go:1119:3: ifElseChain: rewrite if-else to switch statement (gocritic)
		if m.Readonly && m.RecursiveReadOnly {
```

File: server/container_create_linux.go
Comment (kwilczynski) [inline server/container_create_linux.go:996]: **To the reviewer:**

We need to refactor this at some point. The number of arguments is slowly becoming unwieldy.

File: server/container_create_linux.go
Comment (kwilczynski) [inline server/container_create_linux.go:1130]: **To the reviewer:**

When deployed atop Kubernetes, we have the advantage of having the API request validation to tell us that required attributes are not set or have wrong values. However, when creating containers using the API, for example, using crictl, we don't have this added value, and as such, we perform this validation ourselves.

File: server/container_restore.go
Comment (kwilczynski) [inline server/container_restore.go:343]: **To the reviewer:**

Hopefully, this is the right thing to do when restoring a container.

File: pkg/config/config.go
Comment (kwilczynski) [inline pkg/config/config.go:1303]: **To the reviewer:**

Perhaps this should be an `logrus.Infof()`?

Comment (kwilczynski) [reply pkg/config/config.go:1303]: If so, we should add as nice a neat runtime handler features print on startup.

File: pkg/config/config.go
Comment (kwilczynski) [inline pkg/config/config.go:1308]: **To the reviewer:**

An unsupported platform emits:

```console
WARN[2024-04-09 10:25:56.665685980Z] Runtime handler "runc" supports Recursive Read-only (RRO) mounts, but kernel does not: kernel version "5.10.214" does not support recursive read-only mounts: unable to set recursive read-only mount attribute: function not implemented 
WARN[2024-04-09 10:25:56.671210757Z] Runtime handler "crun" supports Recursive Read-only (RRO) mounts, but kernel does not: kernel version "5.10.214" does not support recursive read-only mounts: unable to set recursive read-only mount attribute: function not implemented 
```

File: pkg/config/config.go
Comment (kwilczynski) [inline pkg/config/config.go:1293]: **To the reviewer:**

This was originally missing. I think it was a small bug.

File: server/container_create_linux.go
Comment (haircommander) [reply server/container_create_linux.go:996]: very true. @hasan4791 is working on a refactor which could handle this

File: server/container_create_linux.go
Comment (haircommander) [reply server/container_create_linux.go:1129]: I kinda feel we should do this validation in crictl to match kube doing it but I am open to double validating

File: server/container_create_linux.go
Comment (haircommander) [inline server/container_create_linux.go:1139]: why not another case here?

File: server/container_restore.go
Comment (kwilczynski) [inline server/container_restore.go:342]: **To the reviewer:**

Hopefully, this is the right thing to do when restoring a container.

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go:1129]: It depends on how you look at it. If you take CRI-O as an API server, then providing server-side validation, so to speak, would be more than fine, like the 412 HTTP status code in RESTful endpoints.

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go:1139]: Sure.

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go:996]: Thank you, @hasan4791, in advance! :tada:

File: server/container_create_linux.go
Comment (kwilczynski) [reply server/container_create_linux.go]: Done.

**Comment by haircommander** [comment]
/approve

LGTM, thanks! @cri-o/cri-o-maintainers PTAL

File: pkg/config/config_linux.go
Comment (saschagrunert) [inline pkg/config/config_linux.go:51]: nit: we can use `errors.Join` here.

**Comment by kwilczynski** [comment]
/hold

For a brief moment.

File: pkg/config/config_linux.go
Comment (kwilczynski) [reply pkg/config/config_linux.go:51]: The problem with `errors.Join()` is this:

```
// The error formats as the concatenation of the strings obtained
// by calling the Error method of each element of errs, with a newline
// between each string.
```

This sadly breaks logrus per:

```console
WARN[2024-04-11 08:03:38.207720457Z] Runtime handler "runc" supports Recursive Read-only (RRO) mounts, but kernel does not: kernel version "5.10.214" does not support recursive read-only mounts
unable to set recursive read-only mount attribute: function not implemented
WARN[2024-04-11 08:03:38.215434604Z] Runtime handler "crun" supports Recursive Read-only (RRO) mounts, but kernel does not: kernel version "5.10.214" does not support recursive read-only mounts
unable to set recursive read-only mount attribute: function not implemented
```

This is an unfortunate API that Go has that does not allow you to customise the `Error()` method of the type that handles joining errors.

I can replace `%v` with `%w`, as from 1.20 onwards, `fmt.Errorf()` supports multiple error wrapping. This will have the same effect, sans the better error unwrapping if needed.

File: test/ctr.bats
Comment (kwilczynski) [inline test/ctr.bats:1032]: @littlejawa, do you happen to know off the cuff, perhaps, whether any read-only mounts under Kata are also recursively read-only for the entire tree? So, would a sub-volume mount also be read-only?

I am asking, as this simple above test does not pass under Kata, whereas it has no issues on a vanilla tux container.

Comment (kwilczynski) [reply test/ctr.bats:1032]: Probably not. I am getting the same failure on Fedora.

File: contrib/test/ci/vars.yml
Comment (kwilczynski) [inline contrib/test/ci/vars.yml:136]: **To the reviewer:**

I am not sure if Kata supports this correctly. I couldn't get it to work. Perhaps it's the [virtio-fs](https://virtio-fs.gitlab.io/) that's behind the scenes that requires some setting to be correctly enabled.

**Comment by haircommander** [comment]
/approve

if there's another round of reviews, refactoring the commits to break up the test changes `supports_selinux` pieces and the addition of the code would be nice, but nothing to block on IMO

@cri-o/cri-o-maintainers PTAL

**Comment by kwilczynski** [comment]
> if there's another round of reviews, refactoring the commits to break up the test changes `supports_selinux` pieces and the addition of the code would be nice, but nothing to block on IMO

@haircommander, done.

File: pkg/config/config.go
Comment (kolyshkin) [reply pkg/config/config.go:1317]: If we're ignoring the error here, should it be `logrus.Warnf` not `Errorf`?

File: pkg/config/config.go
Comment (kolyshkin) [inline pkg/config/config.go:72]: AFAICS this needs to be moved to config_linux.go, as it's only used from there.

File: pkg/config/config_linux.go
Comment (kolyshkin) [inline pkg/config/config_linux.go:91]: Maybe it's easier to use github.com/containers/storage/pkg/parsers/kernel rather than reimplement uname parsing here?

File: pkg/config/config_linux.go
Comment (kolyshkin) [inline pkg/config/config_linux.go:103]: nit: os.RemoveAll error is like "rmdir %s: %w" (meaning you don't have to include the directory name and the fact that directory removal failed, as it's already in the error).

File: pkg/config/config.go
Comment (kwilczynski) [reply pkg/config/config.go:72]: The idea was to share this amongst other implementations. However, I don't know if we will ever get FreeBSD one to work with. I will move it, indeed.

File: pkg/config/config_linux.go
Comment (kwilczynski) [reply pkg/config/config_linux.go:91]: I had looked at this package before; it was quite elaborate, given what we needed here. However, I will switch to it to benefit from a much more widely used and better-tested implementation. 👍

File: pkg/config/config_linux.go
Comment (kwilczynski) [reply pkg/config/config_linux.go:103]: Ah, yes. The `PathError` type captures it. Will change.

File: pkg/config/config.go
Comment (kwilczynski) [reply pkg/config/config.go]: Done.

File: pkg/config/config_linux.go
Comment (kwilczynski) [reply pkg/config/config_linux.go]: Done.

File: pkg/config/config_linux.go
Comment (kwilczynski) [reply pkg/config/config_linux.go]: Done.

File: pkg/config/config.go
Comment (kwilczynski) [reply pkg/config/config.go:1311]: I can change these to warnings. However, an error carries much more weight than a warning when we want to bring something to the user's attention, especially during initialisation. Perhaps that is why this was originally done that way.

File: contrib/test/ci/vars.yml
Comment (littlejawa) [reply contrib/test/ci/vars.yml:136]: I can't answer from the top of my head. It's probably better to keep them in the skip list for now.
Thanks!

**Comment by kwilczynski** [comment]
@cri-o/cri-o-maintainers, please have a look again. Thank you!

File: pkg/config/config.go
Comment (kwilczynski) [reply pkg/config/config.go:1329]: The output format remains the same even after following the refactor as suggested in the review comment:

- https://github.com/cri-o/cri-o/pull/7962#pullrequestreview-2002056079

**Review by sohankunkerkar — APPROVED** [review]
/lgtm

File: test/ctr.bats
Comment (kwilczynski) [reply test/ctr.bats:1012]: @sohankunkerkar, we could.

However, this specific set of `tmpfs` mount points is only ever going to be used in these two tests. I am unsure if abstracting it out adds the benefits of reusability here. We need it to be done specifically this way, and it's not applicable outside these two tests, in other words.

Thoughts?

Comment (sohankunkerkar) [reply test/ctr.bats:1012]: Yeah, I agree with you. I was thinking if we could do it in the same test, but feel free to ignore this suggestion if it doesn't make sense.

Comment (kwilczynski) [reply test/ctr.bats]: No worries. This is now done.

**Comment by kwilczynski** [comment]
OK. Requires manual cherry-pick.

**Comment by saschagrunert** [comment]
> OK. Requires manual cherry-pick.

No need to cherry-pick this one as discussed.

**Comment by kwilczynski** [comment]
> > OK. Requires manual cherry-pick.
> 
> No need to cherry-pick this one as discussed.

To add more background here:

I initially thought we would need to backport this to 1.29, to support older Kubernetes and, as such, OpenShift. Thinking that we were missing some crucial functionality; however, this feature, as I remembered later, is only made available as Alpha in Kubernetes 1.30. Thus, there is no point in having any backports.

**Pull request merged**
